### PR TITLE
Handle page select and remote access check after changing SFP target

### PR DESCRIPTION
**Pull request by mihirpat1**
#### Description

After changing the target to remote, many sfputil commands which access transceiver EEPROM fail with the below traceback if either of the below scenario is true.
1. If the remote target had page select register set to non-zero
2. If the remote target was powered down or is removed from the device

Traceback details
```
Running command: sudo sfputil show fwversion Ethernet232
['Traceback (most recent call last):', '  File "/usr/local/bin/sfputil", line 8, in <module>', '    sys.exit(cli())', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 764, in __call__', '    return self.main(*args, **kwargs)', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 717, in main', '    rv = self.invoke(ctx)', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1137, in invoke', '    return _process_result(sub_ctx.command.invoke(sub_ctx))', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1137, in invoke', '    return _process_result(sub_ctx.command.invoke(sub_ctx))', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 956, in invoke', '    return ctx.invoke(self.callback, **ctx.params)', '  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 555, in invoke', '    return callback(*args, **kwargs)', '  File "/usr/local/lib/python3.9/dist-packages/sfputil/main.py", line 1056, in fwversion', '    show_firmware_version(physical_port)', '  File "/usr/local/lib/python3.9/dist-packages/sfputil/main.py", line 1026, in show_firmware_version', '    api = sfp.get_xcvr_api()', '  File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sfp_base.py", line 468, in get_xcvr_api', '    self.refresh_xcvr_api()', '  File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sfp_base.py", line 458, in refresh_xcvr_api', '    self._xcvr_api = self._xcvr_api_factory.create_xcvr_api()', '  File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/xcvr_api_factory.py", line 76, in create_xcvr_api', '    vendor_pn = self._get_vendor_part_num()', '  File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/xcvr_api_factory.py", line 67, in _get_vendor_part_num', '    vendor_pn = part_num.decode()', "UnicodeDecodeError: 'utf-8' codec can't decode byte 0x9d in position 2: invalid start byte"]
```

#### Motivation and Context

The set_firmware_download_target_end API needs to be enhanced since it currently
1. Does not set the page select byte to 0 after changing the target (assumes it to be zero). In addition, the optoe kernel driver changes page select byte only if a non-zero page is being accessed
https://github.com/opencomputeproject/OpenNetworkLinux/blob/e85f118fb6400caf869f4a83b92c0bc3123962da/packages/base/any/kernels/modules/optoe.c#L522
2. Does not ensure if the remote target is accessible after changing the target.

If either of the above scenario is true, when sfputil or xcvrd attempts reading the Vendor name and Vendor part number (residing an Upper page 0h) to create the relevant transceiver specific instances (https://github.com/sonic-net/sonic-platform-common/blob/de16e50d87f63fb3ab3cfa9f52c8d2557de67ebb/sonic_platform_base/sonic_xcvr/xcvr_api_factory.py#L72), an unexpected data is read since
1. If the current scenario is `#1`, the data of upper page will be actually read from the corresponding non-zero page selected
2. If the current scenarios is `#2`, all registers of EEPROM except TARGET_MODE register will have a value of 0xff populated

In addition to the above changes, the `get_current_target_end` API has been added for future use.

#### How Has This Been Tested?

Following test cases were executed

1. Change target to E1 and dump page 1
2. Modify page select to 0x10 on E1 and change target to E1
3. Modify page select to 0x11 on E2 and change target to E2
4. Remove E1, change target to E1 and ensure target is changed back to 0
5. Ensure firmware info is not populated when remote target is not accessible
6. Ensure get_firmware_download_target_end returns valid output 

#### Additional Information (Optional)
MSFT ADO - 27920899

### CODE REVIEW

**Comment by prgeor** [comment]
@mihirpat1 you test the case where 
1. remote target is powered off or removed

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py [hunk 1]
```diff
@@ -97,6 +120,31 @@ def get_transceiver_info_firmware_versions(self):
...
+        Returns:
+            True if the remote target is accessible from the local target, False otherwise.
+        """
+        module_type = self.get_module_type()
+        if 'Unknown' in module_type:
+            return False
+
+        return True
+
+    def _handle_error_and_restore_target_to_E0(self, error_message):
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py:138]: @mihirpat1 this API be rename to _restore_target_E0(). Let the caller print the log errors before calling this. This allows reusing this API in future without throwing error logs

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py [hunk 2]
```diff
@@ -44,7 +43,38 @@
...
+                if not self._is_remote_target_accessible():
+                    return self._handle_error_and_restore_target_to_E0(
+                            "Remote target {} not accessible.".format(target))
+        except Exception as e:
+            return self._handle_error_and_restore_target_to_E0(
+                    "Exception occurred while setting target mode to {}: {}".format(target, repr(e)))
+
+        return True
+
+    def get_firmware_download_target_end(self):
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py:71]: @mihirpat1 nit, rename to get_current_target_end()

Comment (prgeor) [reply sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py:71]: @mihirpat1 please check if this is tested for
1. remote target is already powered off
2. remote target is powered off after setting the target > 0

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py (hunk 1 above)
Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py]: @prgeor I have addressed this now.

File: sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py (hunk 2 above)
Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmisTargetFWUpgrade.py]: @prgeor I have addressed this now. I haven't tested the power off scenario though.

**Comment by mihirpat1** [comment]
> @mihirpat1 you test the case where
> 
> 1. remote target is powered off or removed

@prgeor I have tested the case wherein remote target is removed

**Review by prgeor — APPROVED** [review]

**Comment by StormLiangMS** [comment]
hi @mihirpat1  @prgeor there is a conflict, could you raise separate PR?

**Comment by mihirpat1** [comment]
> hi @mihirpat1 @prgeor there is a conflict, could you raise separate PR?

@StormLiangMS Can you please try to cherry pick now since the dependent PR is merged?

**Comment by StormLiangMS** [comment]
could you update the test result for 202311 @mihirpat1 ?

**Comment by mihirpat1** [comment]
> could you update the test result for 202311 @mihirpat1 ?

@StormLiangMS I have now tested this on 202311. Can you please help to merge this?

**Pull request merged**